src/test/java/com/yelp/nrtsearch/server/remote/s3/S3BackendTest.java

```diff
@@ -0,0 +1,390 @@
+    String latestKey =
+        S3Backend.getVersionKey("download_service_3", resource, S3Backend.LATEST_VERSION);
+    s3.putObject(BUCKET_NAME, latestKey, "1");
+
+    try {
+      s3Backend.downloadStream(
+          "download_service_3", "download_index_3", IndexResourceType.WARMING_QUERIES);
+      fail();
+    } catch (IllegalArgumentException e) {
+      assertEquals(
+          "Object s3://s3-backend-test/download_service_3/_version/download_index_3_warming_queries/1 not found",
+          e.getMessage());
+    }
+
+    String versionKey = S3Backend.getVersionKey("download_service_3", resource, "1");
+    String resourceHash = UUID.randomUUID().toString();
+    s3.putObject(BUCKET_NAME, versionKey, resourceHash);
+
+    try {
+      s3Backend.downloadStream(
+          "download_service_3", "download_index_3", IndexResourceType.WARMING_QUERIES);
+      fail();
+    } catch (IllegalArgumentException e) {
+      assertEquals(
+          "Object s3://s3-backend-test/download_service_3/download_index_3_warming_queries/"
+              + resourceHash
+              + " not found",
+          e.getMessage());
+    }
+  }
+
+  @Test
+  public void testDownloadStream_updateData() throws IOException {
+    String resource =
+        S3Backend.getResourceName("download_index_4", IndexResourceType.WARMING_QUERIES);
+    String latestKey =
+        S3Backend.getVersionKey("download_service_4", resource, S3Backend.LATEST_VERSION);
+    s3.putObject(BUCKET_NAME, latestKey, "1");
+
+    String versionKey = S3Backend.getVersionKey("download_service_4", resource, "1");
+    String resourceHash = UUID.randomUUID().toString();
+    s3.putObject(BUCKET_NAME, versionKey, resourceHash);
+
+    String resourceKey = S3Backend.getResourceKey("download_service_4", resource, resourceHash);
+    s3.putObject(BUCKET_NAME, resourceKey, "resource_data_1");
+
+    InputStream downloadStream =
+        s3Backend.downloadStream(
+            "download_service_4", "download_index_4", IndexResourceType.WARMING_QUERIES);
+    assertEquals("resource_data_1", convertToString(downloadStream));
+
+    // update resource data
+    String resourceHash2 = UUID.randomUUID().toString();
+    String resourceKey2 = S3Backend.getResourceKey("download_service_4", resource, resourceHash2);
+    s3.putObject(BUCKET_NAME, resourceKey2, "resource_data_2");
+
+    String versionKey2 = S3Backend.getVersionKey("download_service_4", resource, "2");
+    s3.putObject(BUCKET_NAME, versionKey2, resourceHash2);
+    s3.putObject(BUCKET_NAME, latestKey, "2");
+
+    downloadStream =
+        s3Backend.downloadStream(
+            "download_service_4", "download_index_4", IndexResourceType.WARMING_QUERIES);
+    assertEquals("resource_data_2", convertToString(downloadStream));
+  }
+
+  @Test
+  public void testUploadFile() throws IOException {
+    File uploadFile = S3_PROVIDER.getTemporaryFolder().newFile("upload_file");
```

Comment:
It seems weird to get TemporaryFolder from S3Provider and use it. Unless someone reads the code, they might just put things in the s3 directory via the TemporaryFolder without meaning to. Maybe just create TemporaryFolder here for this class and remove the getter?